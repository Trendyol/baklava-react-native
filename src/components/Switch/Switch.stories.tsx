import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react-native';
import Switch from './Switch';
import theme from '../../theme';
import Text from '../Text/Text';
import Box from '../Box/Box';
import Button from '../Button/Button';

const variantList = Object.keys(theme.textVariants);

const TextMeta: ComponentMeta<typeof Switch> = {
  title: 'Switch',
  component: Switch,
  args: {
    label: variantList[0] as string,
  },
};

export default TextMeta;

type SwitchStory = ComponentStory<typeof Switch>;

export const Basic: SwitchStory = () => (
  <>
    <Text p="2xs" variant="subtitle01Bold">
      Switch
    </Text>

    <Box px="m" py="2xs">
      <Switch onValueChange={_ => {}} />
    </Box>
  </>
);

export const SwitchWithLabel: SwitchStory = args => (
  <>
    <Text p="2xs" variant="subtitle01Bold">
      Switch With Label
    </Text>

    <Box px="m" py="2xs">
      <Switch onValueChange={_ => {}} label={args.label} disabled={false} />
    </Box>
  </>
);

export const DisabledSwitch: SwitchStory = () => (
  <>
    <Text p="2xs" variant="subtitle01Bold">
      Disabled Switch
    </Text>

    <Box px="m" py="2xs">
      <Switch onValueChange={_ => {}} disabled={true} />
    </Box>

    <Box px="m" py="2xs">
      <Switch onValueChange={_ => {}} disabled={true} value={true} />
    </Box>
  </>
);

export const CustomSwitch: SwitchStory = () => (
  <>
    <Text p="2xs" variant="subtitle01Bold">
      Custom Switch
    </Text>

    <Box px="m" py="2xs">
      <Switch
        onValueChange={_ => {}}
        onColor={theme.colors.successKey}
        offColor={theme.colors.dangerKey}
      />
    </Box>
  </>
);

export const SwitchState: SwitchStory = () => {
  const [value, setValue] = React.useState(true);

  return (
    <>
      <Text p="2xs" variant="subtitle01Bold">
        Switch State
      </Text>

      <Box px="m" py="2xs" flexDirection="row">
        <Box pr="3xl">
          <Switch onValueChange={setValue} value={value} label="Lorem" />
        </Box>

        <Button
          label={'Make it ' + !value}
          onPress={() => {
            setValue(!value);
          }}
        />
      </Box>
    </>
  );
};
