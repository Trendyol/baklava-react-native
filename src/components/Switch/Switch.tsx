import {
  createRestyleComponent,
  layout,
  border,
  spacing,
  opacity,
} from '@ergenekonyigit/restyle';
import React, { useRef, useState } from 'react';
import {
  SwitchProps as RNSwitchProps,
  Animated,
  Easing,
  Pressable,
  StyleProp,
  ViewStyle,
} from 'react-native';
import theme, { Theme } from '../../theme';
import Text from '../Text/Text';
import Box from '../Box/Box';

export type SwitchContainerProps = React.ComponentProps<typeof Box>;

const SwitchContainer = createRestyleComponent<SwitchContainerProps, Theme>(
  [layout],
  Box,
);

export type AnimatedViewProps = React.ComponentProps<typeof Box> &
  Animated.AnimatedProps<StyleProp<ViewStyle>>;

const AnimatedView = createRestyleComponent<AnimatedViewProps, Theme>(
  [layout, border, spacing],
  Animated.View,
);

export type PressableProps = React.ComponentProps<typeof Box> &
  React.ComponentProps<typeof Pressable>;

const RestyledPressable = createRestyleComponent<PressableProps, Theme>(
  [opacity],
  Pressable,
);

export type SwitchProps = RNSwitchProps & {
  label?: string;
  value?: boolean;
  onValueChange: (value: boolean) => void;
  disabled?: boolean;
  onColor?: string;
  offColor?: string;
  containerStyle?: StyleProp<ViewStyle>;
};

const Switch = ({
  label,
  value,
  onValueChange,
  disabled,
  onColor,
  offColor,
  testID,
  accessibilityLabel,
  accessible,
  containerStyle,
}: SwitchProps) => {
  const [isEnabled, setIsEnabled] = useState(value ? value : false);
  const animatedCirclePosition = useRef(
    new Animated.Value(isEnabled ? 1 : 0),
  ).current;
  const animatedSwitchWidth = useRef(
    new Animated.Value(theme.spacing.l),
  ).current;
  const animatedSwitchMarginRight = useRef(new Animated.Value(0)).current;

  const onSwitchPressIn = () => {
    Animated.timing(animatedSwitchMarginRight, {
      toValue: isEnabled ? 1 : -1,
      duration: 200,
      easing: Easing.sin,
      useNativeDriver: false,
    }).start();

    Animated.timing(animatedSwitchWidth, {
      toValue: theme.spacing.xl,
      duration: 200,
      easing: Easing.sin,
      useNativeDriver: false,
    }).start();
  };

  const onSwitchPressOut = () => {
    Animated.timing(animatedSwitchWidth, {
      toValue: theme.spacing.l,
      duration: 130,
      easing: Easing.out(Easing.exp),
      useNativeDriver: false,
    }).start();
  };

  const handleValueChange = React.useCallback(
    (nextState: boolean) => {
      setIsEnabled(nextState);
      if (onValueChange) {
        onValueChange(nextState);
      }

      Animated.timing(animatedCirclePosition, {
        toValue: nextState ? 1 : 0,
        duration: 130,
        easing: Easing.out(Easing.exp),
        useNativeDriver: false,
      }).start();

      Animated.timing(animatedSwitchMarginRight, {
        toValue: 0,
        duration: 90,
        easing: Easing.out(Easing.exp),
        useNativeDriver: false,
      }).start();
    },
    [animatedCirclePosition, animatedSwitchMarginRight, onValueChange],
  );

  React.useEffect(() => {
    handleValueChange(!!value);
  }, [value, handleValueChange]);

  const testProps = React.useMemo(() => {
    const result: RNSwitchProps = {
      accessible: false,
      accessibilityLabel: undefined,
      testID: undefined,
    };

    if (!(testID || accessibilityLabel)) {
      return result;
    }

    result.accessible = accessible ?? true;
    result.accessibilityLabel = accessibilityLabel;
    result.testID = testID ?? accessibilityLabel;

    return result;
  }, [testID, accessibilityLabel, accessible]);

  return (
    <SwitchContainer
      flexDirection="row"
      alignItems="center"
      style={containerStyle}>
      {label && (
        <Text variant="subtitle2Medium" p="2xs" testID="switch-label">
          {label}
        </Text>
      )}
      <RestyledPressable
        opacity={disabled ? 0.5 : 1}
        disabled={disabled}
        onPressOut={() => onSwitchPressOut()}
        onPressIn={() => onSwitchPressIn()}
        onPress={() => handleValueChange(!isEnabled)}
        {...testProps}>
        <AnimatedView
          width={theme.spacing['4xl']}
          height={theme.spacing.xl}
          justifyContent="center"
          borderRadius="full"
          testID="switch-outer-view"
          //@ts-ignore
          backgroundColor={animatedCirclePosition.interpolate({
            inputRange: [0, 1],
            outputRange: [
              offColor ? offColor : theme.colors.neutralLighter,
              onColor ? onColor : theme.colors.primaryKey,
            ],
          })}>
          <AnimatedView
            backgroundColor="white"
            height={theme.spacing.l}
            marginStart="4xs"
            borderRadius="full"
            testID="switch-inner-view"
            //@ts-ignore
            width={animatedSwitchWidth}
            //@ts-ignore
            style={{
              transform: [
                {
                  translateX: animatedCirclePosition.interpolate({
                    inputRange: [0, 1],
                    outputRange: [0, theme.spacing.xl],
                  }),
                },
                {
                  translateX: animatedSwitchMarginRight.interpolate({
                    inputRange: [-1, 0, 1],
                    outputRange: [2, 0, -6],
                  }),
                },
              ],
            }}
          />
        </AnimatedView>
      </RestyledPressable>
    </SwitchContainer>
  );
};

export default Switch;
