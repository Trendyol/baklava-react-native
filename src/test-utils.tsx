import { ThemeProvider } from '@ergenekonyigit/restyle';
import { render } from '@testing-library/react-native';
import * as React from 'react';
import {
  GestureResponderHandlers,
  PanResponder,
  PanResponderCallbacks,
} from 'react-native';
import theme from './theme';
import { PortalProvider } from '.';

theme.fonts = {
  light: 'Rubik-Light',
  regular: 'Rubik-Regular',
  medium: 'Rubik-Medium',
  semiBold: 'Rubik-SemiBold',
  bold: 'Rubik-Bold',
};

export interface RenderOptions {
  wrapper?: React.ComponentType<any>;
  createNodeMock?: (element: React.ReactElement<any>) => any;
}

const AllProviders = ({ children }: { children?: React.ReactNode }) => (
  <ThemeProvider theme={theme}>
    <PortalProvider>{children}</PortalProvider>
  </ThemeProvider>
);

const customRender = (ui: React.ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: AllProviders, ...options }) as any;

const mockPanResponder = () => {
  jest
    .spyOn(PanResponder, 'create')
    .mockImplementation(
      ({
        onStartShouldSetPanResponderCapture,
        onMoveShouldSetPanResponder,
        onPanResponderGrant,
        onPanResponderMove,
        onPanResponderRelease,
      }: PanResponderCallbacks) => ({
        panHandlers: {
          onStartShouldSetResponderCapture: onStartShouldSetPanResponderCapture,
          onMoveShouldSetResponder: onMoveShouldSetPanResponder,
          onResponderGrant: onPanResponderGrant,
          onResponderMove: onPanResponderMove,
          onResponderRelease: onPanResponderRelease,
        } as GestureResponderHandlers,
      }),
    );
};

export * from '@testing-library/react-native';
export { customRender as render, mockPanResponder };
