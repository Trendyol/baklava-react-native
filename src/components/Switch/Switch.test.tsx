import React from 'react';
import Switch from './Switch';
import { render } from '../../test-utils';
import Theme from '../../theme';

describe('Switch', () => {
  test('should render switch correctly', () => {
    // when
    const { toJSON } = render(
      <Switch onValueChange={() => {}} accessibilityLabel={'switch'} />,
    );

    // then
    expect(toJSON()).toMatchSnapshot();
  });

  test('should render disabled switch correctly', () => {
    // when
    const { getByTestId } = render(
      <Switch onValueChange={() => {}} testID="switch" disabled={true} />,
    );
    const switchComponent = getByTestId('switch');

    // then
    expect(switchComponent.props.style[0].opacity).toBe(0.5);
  });

  test('should render disabled switch with value true correctly', () => {
    // when
    const { getByTestId } = render(
      <Switch
        onValueChange={() => {}}
        testID="switch"
        value={true}
        disabled={true}
      />,
    );
    const switchComponent = getByTestId('switch');
    const switchOuterView = getByTestId('switch-outer-view');

    // then
    expect(switchComponent.props.style[0].opacity).toBe(0.5);
    expect(switchOuterView._fiber.pendingProps.backgroundColor).toBe(
      'rgba(242, 122, 26, 1)',
    );
  });

  test('should render disabled switch with value false correctly', () => {
    // when
    const { getByTestId } = render(
      <Switch
        onValueChange={() => {}}
        testID="switch"
        value={false}
        disabled={true}
        onColor={undefined}
        offColor={undefined}
      />,
    );
    const switchComponent = getByTestId('switch');
    const switchOuterView = getByTestId('switch-outer-view');

    // then
    expect(switchOuterView._fiber.pendingProps.backgroundColor).toBe(
      'rgba(175, 187, 202, 1)',
    );
    expect(switchComponent.props.style[0].opacity).toBe(0.5);
  });

  test('should render given switch label correctly', () => {
    // when
    const { getByTestId } = render(
      <Switch
        onValueChange={() => {}}
        testID={'switch'}
        label={'test switch label'}
      />,
    );
    const switchComponent = getByTestId('switch-label');

    // then
    expect(switchComponent.props.children).toBe('test switch label');
  });

  test('should render customized switch correctly', () => {
    // when
    const { getByTestId } = render(
      <Switch
        onValueChange={() => {}}
        testID={'switch'}
        label={'test switch label'}
        onColor={Theme.colors.successColor}
        offColor={Theme.colors.dangerColor}
      />,
    );
    const switchOuterView = getByTestId('switch-outer-view');

    //then
    expect(switchOuterView._fiber.pendingProps.backgroundColor).toBe(
      'rgba(255, 80, 67, 1)',
    );
  });
});
