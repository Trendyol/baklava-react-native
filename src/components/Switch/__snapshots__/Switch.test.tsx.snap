// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch should render switch correctly 1`] = `
<View
  accessible={false}
  style={
    Array [
      Object {},
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        },
      ],
    ]
  }
>
  <View
    accessibilityLabel="switch"
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "opacity": 1,
        },
      ]
    }
    testID="switch"
  >
    <View
      accessibilityLabel="switch-outer-view"
      backgroundColor="rgba(175, 187, 202, 1)"
      collapsable={false}
      style={
        Object {
          "alignItems": "center",
          "borderRadius": 999,
          "flexDirection": "row",
          "height": 24,
          "paddingHorizontal": 2,
          "width": 48,
        }
      }
      testID="switch-outer-view"
    >
      <View
        accessibilityLabel="switch-inner-view"
        backgroundColor="white"
        collapsable={false}
        style={
          Object {
            "borderRadius": 999,
            "height": 20,
            "transform": Array [
              Object {
                "translateX": 0,
              },
              Object {
                "translateX": 0,
              },
            ],
            "width": 20,
          }
        }
        testID="switch-inner-view"
      />
    </View>
  </View>
</View>
`;
